File: crossref_diet_products.js
/**
 * crossref_diet_products.js
 * ──────────────────────────
 * For each uncertain diet flag:
 *   1. Uses Claude to extract the specific problematic ingredient from the reason text
 *   2. Searches all 329k FIG products for that ingredient by name
 *   3. Checks protocol compliance (compliant / caution / not_compliant)
 *
 * Rules:
 *   - compliant product found  → mod confirmed
 *   - caution only             → grey area (noted)
 *   - not_compliant only       → mod not possible
 *   - nothing found            → no product in FIG DB
 *   - identity-destroying list → mod: false, skip search
 *
 * Keto uses sugar_free + paleo (both must be compliant)
 * Caution = not compliant unless a compliant product also exists
 *
 * Usage:
 *   node crossref_diet_products.js
 *   node crossref_diet_products.js --concurrency 5
 */

const Anthropic = require('./functions/node_modules/@anthropic-ai/sdk');
const fs        = require('fs');
const path      = require('path');

// ── Config ─────────────────────────────────────────────────────────────────────
const UNCERT_FILE   = path.join(__dirname, 'diet_uncertainty_report.json');
const PRODUCTS_FILE = '/Users/rafi/Desktop/Claude-MHC/Fig Scraper/ckc_products_cleaned_2026-03-29.json';
const PROGRESS_FILE = path.join(__dirname, 'crossref_progress.json');
const OUT_FILE      = path.join(__dirname, 'diet_product_crossref_report.json');

const envContent  = fs.readFileSync(path.join(__dirname, 'functions', '.env'), 'utf8');
const apiKeyMatch = envContent.match(/ANTHROPIC_API_KEY=(.+)/);
if (!apiKeyMatch) { console.error('ANTHROPIC_API_KEY not found'); process.exit(1); }
process.env.ANTHROPIC_API_KEY = apiKeyMatch[1].trim();

const args        = process.argv.slice(2);
const CONCURRENCY = parseInt(args.find(a => a.startsWith('--concurrency='))?.split('=')[1] || '5');
const RESET       = args.includes('--reset');

const client = new Anthropic();

// ── Protocol → FIG field ───────────────────────────────────────────────────────
const PROTO_FIELD = {
  AIP: 'aip_friendly',
  LF:  'low_fodmap',
  GF:  'gluten_free',
  DF:  'dairy_free',
  Vg:  'vegan',
  V:   'vegetarian',
  LH:  'low_histamine',
  K:   null, // special: sugar_free + paleo
};

// ── Identity-destroying → mod: false, skip search ─────────────────────────────
const IDENTITY_DESTROYING = new Set([
  'K|Crispy Falafel Recipe',
  'K|Mediterranean Lentil Salad',
  'LF|Peanut Butter Chicken',
  'LH|Guacamole',
  'LH|Mango Pico de Gallo',
  'LF|Double the Mushrooms Chicken Marsala',
  'LH|Salmon Puttanesca',
  'LH|Castelvetrano Olive Chicken Skillet',
  'LH|Creamy Refried Beans',
  'K|Lightened Sweet Potato Casserole Pecan Oat Streusel',
  'LH|Fettuccine with Smoked Salmon and Dill Cream Sauce',
  'LH|Miso Glazed Salmon Bowls',
  'LH|Tomato Aguachile',
]);

// ── Progress ───────────────────────────────────────────────────────────────────
function loadProgress() {
  if (!RESET && fs.existsSync(PROGRESS_FILE)) {
    return JSON.parse(fs.readFileSync(PROGRESS_FILE, 'utf8'));
  }
  return { results: [] };
}

function saveProgress(progress) {
  fs.writeFileSync(PROGRESS_FILE, JSON.stringify(progress, null, 2));
}

// ── Product compliance check ───────────────────────────────────────────────────
function getCompliance(product, protocol) {
  if (protocol === 'K') {
    const sf = product.sugar_free || 'unknown';
    const pa = product.paleo      || 'unknown';
    if (sf === 'compliant' && pa === 'compliant') return 'compliant';
    if (sf === 'not_compliant' || pa === 'not_compliant') return 'not_compliant';
    return 'caution';
  }
  const field = PROTO_FIELD[protocol];
  return field ? (product[field] || 'unknown') : 'unknown';
}

function searchProducts(ingredient, protocol, products) {
  const q = ingredient.toLowerCase().trim();
  const results = { compliant: [], caution: [], not_compliant: [] };

  for (const p of products) {
    if (p.name.toLowerCase().includes(q)) {
      const status = getCompliance(p, protocol);
      if (results[status]) {
        results[status].push({ name: p.name, brand: p.brand || '' });
      }
    }
  }
  return results;
}

// ── Claude: extract ingredient from reason ─────────────────────────────────────
async function extractIngredient(recipe, protocol, reason) {
  for (let attempt = 1; attempt <= 3; attempt++) {
    try {
      const resp = await client.messages.create({
        model:      'claude-haiku-4-5-20251001',
        max_tokens: 100,
        system: [
          {
            type: 'text',
            text: `You extract the single most specific problematic ingredient name from a diet compliance uncertainty note.
Return ONLY the ingredient name — 1 to 4 words, lowercase, no punctuation, no explanation.
Examples: "garam masala", "mirin", "taco seasoning", "gochujang", "feta", "scallion", "balsamic vinegar"
If the uncertainty is about serving size or general ambiguity (not a specific ingredient), return: SKIP`,
            cache_control: { type: 'ephemeral' },
          }
        ],
        messages: [
          {
            role: 'user',
            content: `Recipe: ${recipe}\nProtocol: ${protocol}\nReason: ${reason}\n\nIngredient name:`,
          }
        ],
      });

      const text = resp.content[0].text.trim().toLowerCase().replace(/[^a-z0-9\s\-\']/g, '');
      return text === 'skip' ? null : text;

    } catch (err) {
      if (attempt === 3) throw err;
      await new Promise(r => setTimeout(r, attempt * 1000));
    }
  }
}

// ── Sleep ──────────────────────────────────────────────────────────────────────
const sleep = ms => new Promise(r => setTimeout(r, ms));

// ── Process one uncertain item ─────────────────────────────────────────────────
async function processItem(item, index, total, products) {
  const { recipe, protocol, reason, url } = item;
  const key   = `${protocol}|${recipe}`;
  const label = `[${index}/${total}] ${recipe.slice(0, 40).padEnd(40)} ${protocol}`;

  const base = { recipe, protocol, url, reason };

  // Identity-destroying?
  if (IDENTITY_DESTROYING.has(key)) {
    console.log(`${label} → identity-destroying`);
    return { ...base, verdict: 'mod: false — destroys dish identity', category: 'identity_destroying' };
  }

  // Extract ingredient via Claude
  const ingredient = await extractIngredient(recipe, protocol, reason);

  if (!ingredient) {
    console.log(`${label} → SKIP (no specific ingredient)`);
    return { ...base, ingredient_searched: null, verdict: 'skipped — no specific ingredient to search', category: 'skipped' };
  }

  // Search all 329k products
  const matches = searchProducts(ingredient, protocol, products);

  let result;
  if (matches.compliant.length > 0) {
    console.log(`${label} → ✓ compliant (${ingredient})`);
    result = { ...base, ingredient_searched: ingredient, verdict: 'mod confirmed — compliant product exists', compliant_products: matches.compliant.slice(0, 5), category: 'mod_confirmed' };
  } else if (matches.caution.length > 0) {
    console.log(`${label} → ⚠ caution only (${ingredient})`);
    result = { ...base, ingredient_searched: ingredient, verdict: 'grey area — only caution products found', caution_products: matches.caution.slice(0, 5), category: 'grey_area' };
  } else if (matches.not_compliant.length > 0) {
    console.log(`${label} → ✗ not compliant (${ingredient})`);
    result = { ...base, ingredient_searched: ingredient, verdict: 'mod not possible — only non-compliant products found', not_compliant_products: matches.not_compliant.slice(0, 3), category: 'mod_not_possible' };
  } else {
    console.log(`${label} → ? not in FIG DB (${ingredient})`);
    result = { ...base, ingredient_searched: ingredient, verdict: 'no matching product found in FIG database', category: 'no_product_found' };
  }

  return result;
}

// ── Main ───────────────────────────────────────────────────────────────────────
async function main() {
  console.log('CKC Diet × FIG Product Cross-Reference');
  console.log(`Concurrency: ${CONCURRENCY}\n`);

  const uncertain = JSON.parse(fs.readFileSync(UNCERT_FILE, 'utf8'));
  console.log(`Loading ${PRODUCTS_FILE.split('/').pop()}…`);
  const products  = JSON.parse(fs.readFileSync(PRODUCTS_FILE, 'utf8'));
  console.log(`Products loaded: ${products.length.toLocaleString()}\n`);

  const progress = loadProgress();
  const doneKeys = new Set(progress.results.map(r => `${r.protocol}|${r.recipe}`));
  const todo     = uncertain.filter(u => !doneKeys.has(`${u.protocol}|${u.recipe}`));

  console.log(`Total: ${uncertain.length} | Done: ${progress.results.length} | Remaining: ${todo.length}\n`);

  // Process in batches
  for (let i = 0; i < todo.length; i += CONCURRENCY) {
    const batch = todo.slice(i, i + CONCURRENCY);
    const batchResults = await Promise.all(
      batch.map((item, j) => processItem(item, progress.results.length + j + 1, uncertain.length, products))
    );
    progress.results.push(...batchResults);
    saveProgress(progress);
    if (i + CONCURRENCY < todo.length) await sleep(200);
  }

  // Build final report
  const byCategory = {};
  for (const r of progress.results) {
    const cat = r.category || 'unknown';
    if (!byCategory[cat]) byCategory[cat] = [];
    byCategory[cat].push(r);
  }

  const summary = {};
  for (const [cat, items] of Object.entries(byCategory)) {
    summary[cat] = items.length;
  }
  summary.total = progress.results.length;

  const report = { summary, ...byCategory };
  fs.writeFileSync(OUT_FILE, JSON.stringify(report, null, 2));

  console.log('\n── Summary ──────────────────────────────');
  for (const [k, v] of Object.entries(summary)) {
    console.log(`  ${k.padEnd(25)} ${v}`);
  }
  console.log(`\nSaved → ${OUT_FILE}`);
}

main().catch(console.error);

File: crossref_diet_products.py
"""
crossref_diet_products.py
─────────────────────────
Cross-references uncertain diet tag recipes against the FIG product database.

For each uncertain flag:
  - If the recipe modification destroys dish identity → mod: false
  - Otherwise → search FIG products for a compliant version of the flagged ingredient
    - compliant found  → mod confirmed, product noted
    - caution only     → grey area, noted separately
    - nothing found    → mod not possible

Output: diet_product_crossref_report.json
"""

import json, re, os
from collections import defaultdict

# ── Paths ──────────────────────────────────────────────────────────────────────
BASE    = os.path.dirname(os.path.abspath(__file__))
UNCERT  = os.path.join(BASE, 'diet_uncertainty_report.json')
PRODS   = '/Users/rafi/Desktop/Claude-MHC/Fig Scraper/ckc_products_cleaned_2026-03-29.json'
OUT     = os.path.join(BASE, 'diet_product_crossref_report.json')

# ── Protocol → FIG field ───────────────────────────────────────────────────────
PROTO_FIELD = {
    'AIP': 'aip_friendly',
    'LF':  'low_fodmap',
    'GF':  'gluten_free',
    'DF':  'dairy_free',
    'Vg':  'vegan',
    'V':   'vegetarian',
    'LH':  'low_histamine',
    'K':   None,  # special: sugar_free + paleo
}

# ── Identity-destroying cases → mod: false ────────────────────────────────────
IDENTITY_DESTROYING = {
    ('K',  'Crispy Falafel Recipe'),
    ('K',  'Mediterranean Lentil Salad'),
    ('LF', 'Peanut Butter Chicken'),
    ('LH', 'Guacamole'),
    ('LH', 'Mango Pico de Gallo'),
    ('LF', 'Double the Mushrooms Chicken Marsala'),
    ('LH', 'Salmon Puttanesca'),
    ('LH', 'Castelvetrano Olive Chicken Skillet'),
    ('LH', 'Creamy Refried Beans'),
    ('K',  'Lightened Sweet Potato Casserole Pecan Oat Streusel'),
    ('LH', 'Fettuccine with Smoked Salmon and Dill Cream Sauce'),
    ('LH', 'Miso Glazed Salmon Bowls'),
    ('LH', 'Tomato Aguachile'),
}

# ── Ingredient extraction from free-text reason ────────────────────────────────
# Ordered list of regex strategies; first match wins
EXTRACT_PATTERNS = [
    # "X is/are [adjective] [noun]" — subject of sentence
    r'^([A-Za-z][a-zA-Z\s\-\']{2,35}?)\s+(?:is|are)\b',
    # "X composition/content/brand/blend" — strip noise suffix
    r'([A-Za-z][a-zA-Z\s\-\']{2,30}?)\s+(?:composition|content|brand|blend|ingredients?|carb|status)\b',
    # "removing/without/replacing X"
    r'(?:removing|without|replace[sd]?|replacing|swap(?:ping)?)\s+([a-zA-Z][a-zA-Z\s\-\']{2,30}?)(?:\s+(?:is|are|significantly|from|entirely|would|may|can|changes|alters)|\,|\.)',
    # "The X question" / "The X problem"
    r'[Tt]he\s+([a-zA-Z][a-zA-Z\s\-\']{2,25}?)\s+(?:question|problem|issue|concern)\b',
]

# Noise suffixes to strip from extracted terms
NOISE_SUFFIXES = [
    ' composition', ' content', ' brand', ' blend', ' ingredients',
    ' carb', ' status', ' question', ' problem', ' issue',
    ' and ', ' or ',
]

# Non-food stopwords
STOPWORDS = {
    'the', 'this', 'that', 'some', 'most', 'many', 'standard', 'recipe',
    'dish', 'additional', 'context', 'protocol', 'uncertain', 'fodmap',
    'modification', 'access', 'citrus', 'quantity', 'level', 'detail',
    'unknown', 'information', 'amount', 'version', 'option',
}

def extract_ingredient(reason: str) -> list[str]:
    """Return list of ingredient search terms extracted from reason text."""
    candidates = []

    for pat in EXTRACT_PATTERNS:
        m = re.search(pat, reason, re.IGNORECASE)
        if m:
            term = m.group(1).strip().lower()
            candidates.append(term)

    # Clean each candidate
    results = []
    for term in candidates:
        # Strip trailing noise
        for suffix in NOISE_SUFFIXES:
            if term.endswith(suffix.strip()):
                term = term[:term.rfind(suffix.strip())].strip()
        # Strip leading articles
        term = re.sub(r'^(the|a|an)\s+', '', term).strip()
        # Split "X and Y" into two terms
        if ' and ' in term:
            parts = [p.strip() for p in term.split(' and ')]
            results.extend(p for p in parts if len(p) > 2 and p not in STOPWORDS)
        elif len(term) > 2 and term not in STOPWORDS:
            results.append(term)

    return list(dict.fromkeys(results))  # deduplicate, preserve order

# ── Product search ─────────────────────────────────────────────────────────────
def get_compliance(product: dict, protocol: str) -> str:
    if protocol == 'K':
        sf = product.get('sugar_free', 'unknown')
        pa = product.get('paleo', 'unknown')
        if sf == 'compliant' and pa == 'compliant':
            return 'compliant'
        if 'not_compliant' in (sf, pa):
            return 'not_compliant'
        return 'caution'
    field = PROTO_FIELD.get(protocol)
    if not field:
        return 'unknown'
    return product.get(field, 'unknown')

def search_products(query: str, protocol: str, products: list) -> dict:
    q = query.lower()
    results = {'compliant': [], 'caution': [], 'not_compliant': []}
    for p in products:
        if q in p['name'].lower():
            status = get_compliance(p, protocol)
            if status in results:
                results[status].append({'name': p['name'], 'brand': p.get('brand', '')})
    return results

# ── Main ───────────────────────────────────────────────────────────────────────
def main():
    print('Loading uncertainty report…')
    with open(UNCERT) as f:
        uncertain = json.load(f)

    print(f'Loading {PRODS.split("/")[-1]}…')
    with open(PRODS) as f:
        products = json.load(f)
    print(f'Products loaded: {len(products):,}')

    report = {
        'summary': {},
        'identity_destroying':  [],
        'mod_confirmed':        [],
        'grey_area':            [],
        'mod_not_possible':     [],
        'no_product_found':     [],
    }

    totals = defaultdict(int)

    for u in uncertain:
        recipe   = u['recipe']
        protocol = u['protocol']
        reason   = u['reason']
        url      = u.get('url', '')

        base = {
            'recipe':   recipe,
            'protocol': protocol,
            'url':      url,
            'reason':   reason,
        }

        # 1. Identity-destroying?
        if (protocol, recipe) in IDENTITY_DESTROYING:
            report['identity_destroying'].append({**base, 'verdict': 'mod: false — destroys dish identity'})
            totals['identity_destroying'] += 1
            continue

        # 2. Extract ingredient keywords (may return multiple)
        ingredients = extract_ingredient(reason)
        if not ingredients:
            report['no_product_found'].append({**base, 'ingredients_searched': [], 'note': 'Could not extract ingredient keyword'})
            totals['no_extract'] += 1
            continue

        # 3. Search product DB — aggregate across all extracted terms
        all_compliant    = []
        all_caution      = []
        all_not_compliant = []

        for ingredient in ingredients:
            m = search_products(ingredient, protocol, products)
            all_compliant.extend(m['compliant'])
            all_caution.extend(m['caution'])
            all_not_compliant.extend(m['not_compliant'])

        entry = {**base, 'ingredients_searched': ingredients}

        if all_compliant:
            entry['verdict']           = 'mod confirmed — compliant product exists'
            entry['compliant_products'] = all_compliant[:5]
            report['mod_confirmed'].append(entry)
            totals['mod_confirmed'] += 1

        elif all_caution:
            entry['verdict']          = 'grey area — only caution products found'
            entry['caution_products'] = all_caution[:5]
            report['grey_area'].append(entry)
            totals['grey_area'] += 1

        elif all_not_compliant:
            entry['verdict']  = 'mod not possible — only non-compliant products found'
            entry['products'] = all_not_compliant[:3]
            report['mod_not_possible'].append(entry)
            totals['mod_not_possible'] += 1

        else:
            entry['verdict'] = 'no matching product found in FIG database'
            report['no_product_found'].append(entry)
            totals['no_product_found'] += 1

    report['summary'] = {
        'total_uncertain':      len(uncertain),
        'identity_destroying':  totals['identity_destroying'],
        'mod_confirmed':        totals['mod_confirmed'],
        'grey_area':            totals['grey_area'],
        'mod_not_possible':     totals['mod_not_possible'],
        'no_product_found':     totals['no_product_found'] + totals['no_extract'],
    }

    with open(OUT, 'w') as f:
        json.dump(report, f, indent=2)

    print(f'\n── Results ─────────────────────────────')
    for k, v in report['summary'].items():
        print(f'  {k:<25} {v}')
    print(f'\nSaved → {OUT}')

if __name__ == '__main__':
    main()
